runtime/domain.c

```diff
@@ -1099,36 +1109,51 @@ static void install_backup_thread (dom_internal* di)
   sigset_t mask, old_mask;
 #endif
 
-  if (di->backup_thread_running == 0) {
+  if (atomic_load_acquire(&di->backup_thread_running) != 0) {
+    /* If the backup thread is running, but has been instructed to terminate,
+       we need to wait for it to stop until we can spawn another. */
     uintnat msg;
     msg = atomic_load_acquire(&di->backup_thread_msg);
-    CAMLassert (msg == BT_INIT || /* Using fresh domain */
-                msg == BT_TERMINATE); /* Reusing domain */
-
+    if (msg != BT_TERMINATE)
+      return;
     while (msg != BT_INIT) {
       /* Give a chance for backup thread on this domain to terminate */
       caml_plat_unlock (&di->domain_lock);
       cpu_relax ();
       caml_plat_lock_blocking(&di->domain_lock);
       msg = atomic_load_acquire(&di->backup_thread_msg);
     }
+  }
+
+  CAMLassert(atomic_load_acquire(&di->backup_thread_msg) == BT_INIT);
 
 #ifndef _WIN32
-    /* No signals on the backup thread */
-    sigfillset(&mask);
-    pthread_sigmask(SIG_BLOCK, &mask, &old_mask);
+  /* No signals on the backup thread */
+  sigfillset(&mask);
+  pthread_sigmask(SIG_BLOCK, &mask, &old_mask);
 #endif
 
-    atomic_store_release(&di->backup_thread_msg, BT_ENTERING_OCAML);
-    err = pthread_create(&di->backup_thread, 0, backup_thread_func, (void*)di);
-    caml_check_error(err, "failed to create domain backup thread");
+  atomic_store_release(&di->backup_thread_msg, BT_ENTERING_OCAML);
+  err = pthread_create(&di->backup_thread, 0, backup_thread_func, (void*)di);
+  caml_check_error(err, "failed to create domain backup thread");
 
 #ifndef _WIN32
-    pthread_sigmask(SIG_SETMASK, &old_mask, NULL);
+  pthread_sigmask(SIG_SETMASK, &old_mask, NULL);
 #endif
 
-    di->backup_thread_running = 1;
-    pthread_detach(di->backup_thread);
+  atomic_store_release(&di->backup_thread_running, 1);
+  pthread_detach(di->backup_thread);
+}
+
+static void terminate_backup_thread(dom_internal *di)
+{
+  CAMLassert(!caml_bt_is_self());
+
+  if (atomic_load_acquire(&di->backup_thread_running) != 0) {
```

Comment:
```suggestion
  if (atomic_load_acquire(&di->backup_thread_running)) {
```

is more readable? (feel free to ignore).